### ISSUE TITLE
Chat - Composer box validation error!

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22073]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any Chat
2. Add maximum text in composer -> Remove single character and Add Single quote or Double quote

## Expected Result:
 It should allow to send message with single quote or double quote if it’s in text limit

## Actual Result:
Message is not sent if using maximum characters and have single/double quote inside

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.34-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/e18f9d8d-a312-44b2-bfde-014db5ab2a02

https://github.com/Expensify/App/assets/93399543/52d2a048-0661-4c8a-9cf0-e5cefbc6b3ea

**Expensify/Expensify Issue URL:**
**Issue reported by:** @DinalJivani
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688211961584249

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by AdekolaThanni** [comment]
### Proposal

## Please re-state the problem that we are trying to solve in this issue.
Composer box validation error with single and double quotes

## What is the root cause of that problem?
```
// Single quotes - &apos;
// Double quotes - &quot;
```
This is caused by the escaped text being returned after parsing comments greater than CONST.MAX_MARKUP_LENGTH. This converts single quote [ ' ] and double quotes [ " ]to their HTML entities which has been noted above

```
function getParsedComment(text) {
    const parser = new ExpensiMark();
    return text.length < CONST.MAX_MARKUP_LENGTH ? parser.replace(text) : _.escape(text);
}
```

This leads to an increase in the number of characters in the comment that is counted using the ReportsUtils.getCommentLength function which is here 

```
function getCommentLength(textComment) {
    return getParsedComment(textComment)
        .replace(/[^ -~]/g, '\\u????')
        .trim().length;
}
```

## What changes do you think we should make in order to solve the problem?
We could modify the getParsedComment to replace all HTML entities for single quote and double quote before returning

```
function getParsedComment(text) {
    const parser = new ExpensiMark();
    return text.length < CONST.MAX_MARKUP_LENGTH ? parser.replace(text) : _.escape(text).replace(/&apos;/g, " ' ").replace(/&quot;/g, ' " ');
}
```

PS: This will only be a fix for single and double quotes, in any case that we're trying to encompass all other symbols then we'll have no need to escape characters since all symbols will result to an unusual increase in the comment length.

## What alternative solutions did you explore? (Optional)
None

**Comment by AdekolaThanni** [comment]
Sure, I'll be glad to do that.

Email address: adekolathanni@gmail.com
Upwork profile: https://www.upwork.com/freelancers/~0138ee1594d38a0775

**Comment by mallenexpensify** [comment]
@DinalJivani do you have the text/character block you were using to paste in 15000 characters? Asking because I think the issue might be that we're not calculating character counts correctly.

**Comment by AdekolaThanni** [comment]
Exactly @mallenexpensify, just like I said above the character count for a single character which is a symbol gets increased to 6 due to the conversion of that symbol to it's HTML entity form.

**Comment by AdekolaThanni** [comment]
@mallenexpensify, I prepared this test file without any special characters contained. It contains 15000 characters on the dot. I believe you can reproduce the test case with this text file
[test.txt](https://github.com/Expensify/App/files/11966840/test.txt)

**Comment by mallenexpensify** [comment]
Thanks @AdekolaThanni , I missed
> the character count for a single character which is a symbol gets increased to 6 due to the conversion of that symbol to it's HTML entity form

That def makes things _weird_ and inaccurate.   

Do you think the other/better "bug" to fix here would be to treat all characters as single characters so that 15000 is... 15000?  I had a test file I was using and it was ~13000 characters and I was confused.   

I don't think bug/issue is particularly high value, but I'm onboard-enough with having the limit for character count be accurate.

**Comment by AdekolaThanni** [comment]
Yes @mallenexpensify, I believe we should treat all characters as single characters so that 15000 is 15000....

**Comment by AdekolaThanni** [comment]
But it seems the getParsedComment function is used in other places in the codebase, in any case the conversion of the text to its HTML entity form is necessary for the getParsedComment function to work correctly in other places they are used except get CommentLength then I would recommend we add an optional argument to the getParsedComment so it does not break its functionality in other places. This argument could be a condition to specify whether the comment should be escaped on exceeding max length or just returning the text itself would be fine... Do you get my point @mallenexpensify

**Comment by AdekolaThanni** [comment]
But it seems the getParsedComment function is used in other places in the codebase, in any case the conversion of the text to its HTML entity form is necessary for the getParsedComment function to work correctly in other places they are used except get CommentLength then I would recommend we add an optional argument to the getParsedComment so it does not break its functionality in other places. This argument could be a condition to specify whether the comment should be escaped on exceeding max length or just returning the text itself would be fine... Do you get my point @mallenexpensify

**Comment by mallenexpensify** [comment]
I think I do @AdekolaThanni , I'm not an engineer so my knowledge is finite in that realm. 
Looping in @eVoloshchak here for bonus 👀.  Eugene, what do you think a good path forward is here?  There's near-zero value to fixing the bug where 14999 characters and a `"` don't work but I think there might be value is updating and/or standardizing how we acknowledge character counts.

**Comment by eVoloshchak** [comment]
We've had similar issues in https://github.com/Expensify/App/issues/17619 and https://github.com/Expensify/App/issues/18447
To resolve one of those we've had to use mb_strlen on the backend, so that both front end and back end calculate the true number of characters. 
I think that might be applicable here too, but need a bit more time, my initial testing wasn't successful. I'll post an update here tomorrow

**Comment by mallenexpensify** [comment]
Thanks @eVoloshchak , I've removed bug and made this a weekly, since it's not a priority and it's connected/related to other similar issues

**Comment by eVoloshchak** [comment]
To standardize how we acknowledge character counts we will need to

- On the front end we have to make sure we count the true number of characters everywhere (we need to change `getCommentLength` as @AdekolaThanni has proposed). There are also a lot of fields that don't use `getCommentLength`, but still have a length limit (profile, personal info, workspace name inputs). This would require an audit to find all of the affected pages first.
- Count the true number of characters on the back end (since currently, it would still throw an error if we apply FE fix)

I think this requires too many resources for something that adds almost zero actual value to the user experience. We could fix this for this particular case (adding/editing a comment) fairly easily, but fixing it throughout the app on both frontend and backend would be more complex

This needs to be fixed at some point though, but we don't have different priorities we could assign to issues as far as I know

**Comment by mallenexpensify** [comment]
Thanks @eVoloshchak , I agree.  If we don't state anywhere that 15k is the max amount of characters, than nobody would know about this bug either.  I added this to an internal doc I made for 'low value' issues. 
https://docs.google.com/document/d/1S6eiDV3R0oE-UsfC5qae5cOoJZn0TsmprxOqi7htUa0/edit?usp=sharing
